### PR TITLE
cpprest : Fix Eve Online ESI.

### PR DESCRIPTION
**Pull request by p-groarke**
### PR checklist

- [x] Read the [contribution guidelines](https://github.com/swagger-api/swagger-codegen/blob/master/CONTRIBUTING.md).
- [x] Ran the shell script under `./bin/` to update Petstore sample so that CIs can verify the change. (For instance, only need to run `./bin/{LANG}-petstore.sh` and `./bin/security/{LANG}-petstore.sh` if updating the {LANG} (e.g. php, ruby, python, etc) code generator or {LANG} client's mustache templates). Windows batch files can be found in `.\bin\windows\`.
- [x] Filed the PR against the correct branch: `3.0.0` branch for changes related to OpenAPI spec 3.0. Default: `master`.
- [x] Copied the [technical committee](https://github.com/swagger-api/swagger-codegen/#swagger-codegen-technical-committee) to review the pull request if your PR is targeting a particular programming language. @ravinikam (2017/07) @stkrwork (2017/07) @fvarose (2017/11)

### Description of the PR

- Adds a std::vector overload for toJson and toHttpContent functions in ModelBase.
- Fix empty datatypes when response returns "raw" number without enclosing item.
- Fix boost::optional httpBody for POST api.

Cumulative fixes for api : https://esi.tech.ccp.is/latest/swagger.json

https://github.com/swagger-api/swagger-codegen/issues/7100

@etherealjoy As discussed in the issue.

### CODE REVIEW

**Comment by p-groarke** [comment]
Latest commit fixes : https://github.com/swagger-api/swagger-codegen/issues/7572

Renaming to more general intention of PR.

File: modules/swagger-codegen/src/main/resources/cpprest/modelbase-header.mustache [hunk 1]
```diff
@@ -74,6 +80,29 @@ public:
...
+template <class T>
+std::shared_ptr<HttpContent> ModelBase::toHttpContent( const utility::string_t& name, const std::vector<T>& value, const utility::string_t& contentType ) {
+    web::json::value json_array = ModelBase::toJson(value);
+    std::shared_ptr<HttpContent> content( new HttpContent );
+    content->setName( name );
+    content->setContentDisposition( utility::conversions::to_string_t("form-data") );
+    content->setContentType( contentType );
+    std::stringstream* valueAsStringStream = new std::stringstream();
+    (*valueAsStringStream) << json_array;
+    content->setData( std::shared_ptr<std::istream>( valueAsStringStream ) );
```
Comment (etherealjoy) [inline modules/swagger-codegen/src/main/resources/cpprest/modelbase-header.mustache]: Here is missing the encoding conversion

**Comment by etherealjoy** [comment]
@ravinikam @stkrwork @fvarose

File: samples/client/petstore/cpprest/ModelBase.h [hunk 2]
```diff
@@ -86,6 +92,29 @@ class  ModelBase
...
+template <class T>
+std::shared_ptr<HttpContent> ModelBase::toHttpContent( const utility::string_t& name, const std::vector<T>& value, const utility::string_t& contentType ) {
+    web::json::value json_array = ModelBase::toJson(value);
+    std::shared_ptr<HttpContent> content( new HttpContent );
+    content->setName( name );
+    content->setContentDisposition( utility::conversions::to_string_t("form-data") );
+    content->setContentType( contentType );
+    std::stringstream* valueAsStringStream = new std::stringstream();
+    (*valueAsStringStream) << json_array;
+    content->setData( std::shared_ptr<std::istream>( valueAsStringStream ) );
```
Comment (etherealjoy) [inline samples/client/petstore/cpprest/ModelBase.h]: Here is missing the encoding conversion

File: modules/swagger-codegen/src/main/resources/cpprest/api-source.mustache [hunk 3]
```diff
@@ -217,15 +203,11 @@ pplx::task<{{#returnType}}{{{returnType}}}{{/returnType}}{{^returnType}}void{{/r
...
-        }
-        {{/isListContainer}}
-        {{^isListContainer}}
-        {{#isString}}multipart->add(ModelBase::toHttpContent(utility::conversions::to_string_t("{{paramName}}"), {{paramName}}));
-        {{/isString}}
-        {{^isString}}
-        if({{paramName}}.get())
+        }{{/isListContainer}}
+        {{^isListContainer}}{{#isString}}multipart->add(ModelBase::toHttpContent(utility::conversions::to_string_t("{{paramName}}"), {{paramName}}));
+        {{/isString}}{{^isString}}if({{^required}}{{paramName}} && {{/required}}{{paramName}}.get())
```
Comment (etherealjoy) [inline modules/swagger-codegen/src/main/resources/cpprest/api-source.mustache]: if({{^required}}{{paramName}} && {{/required}}{{paramName}}.get())
isn't paramName a pointer here? when ^required?

**Comment by p-groarke** [comment]
A month too late ;)

**Comment by etherealjoy** [comment]
Ahh no problems, I can take your changes and integrate with another PR if you have no objection.

**Comment by p-groarke** [comment]
@etherealjoy Please do! The recursive vector stuff should be useful at minimum. Cheers

**Pull request closed without merging**